src/arkode/arkode_mristep.c

```diff
@@ -1413,239 +1520,1440 @@ int mriStep_FullRHS(ARKodeMem ark_mem, sunrealtype t, N_Vector y, N_Vector f,
+    step_mem->inner_dsm = ZERO;
+    retval = mriStepInnerStepper_ResetAccumulatedError(step_mem->stepper);
+    if (retval != ARK_SUCCESS)
     {
-      N_VConst(ZERO,
-               ark_mem->tempv3); /* set guess to 0 for predictor-corrector form */
-      retval = SUNNonlinSolSetup(step_mem->NLS, ark_mem->tempv3, ark_mem);
-      if (retval < 0) { return (ARK_NLS_SETUP_FAIL); }
-      if (retval > 0) { return (ARK_NLS_SETUP_RECVR); }
+      arkProcessError(ark_mem, ARK_INNERSTEP_FAIL, __LINE__, __func__, __FILE__,
+                      "Unable to reset the inner stepper error estimate");
+      return (ARK_INNERSTEP_FAIL);
     }
-  }
-
-  /* Evaluate the slow RHS functions if needed. NOTE: We do not use the full RHS
-     function here (unlike ERKStep and ARKStep) since it does not need to check
-     for FSAL or SA methods and thus avoids potentially unnecessary evaluations
-     of the inner (fast) RHS function */
-
-  if (!(ark_mem->fn_is_current))
-  {
-    /* compute the explicit component */
-    if (step_mem->explicit_rhs)
+    retval = mriStepInnerStepper_SetRTol(step_mem->stepper,
+                                         step_mem->inner_rtol_factor *
+                                           ark_mem->reltol);
+    if (retval != ARK_SUCCESS)
     {
-      retval = step_mem->fse(ark_mem->tn, ark_mem->yn, step_mem->Fse[0],
-                             ark_mem->user_data);
-      step_mem->nfse++;
-      if (retval) { return ARK_RHSFUNC_FAIL; }
+      arkProcessError(ark_mem, ARK_INNERSTEP_FAIL, __LINE__, __func__, __FILE__,
+                      "Unable to set the inner stepper tolerance");
+      return (ARK_INNERSTEP_FAIL);
     }
+  }
 
-    /* compute the implicit component */
-    if (step_mem->implicit_rhs)
+  /* for adaptive computations, reset the inner integrator to the beginning of this step */
+  if (!ark_mem->fixedstep)
+  {
+    retval = mriStepInnerStepper_Reset(step_mem->stepper, t0, ark_mem->yn);
+    if (retval != ARK_SUCCESS)
     {
-      retval = step_mem->fsi(ark_mem->tn, ark_mem->yn, step_mem->Fsi[0],
-                             ark_mem->user_data);
-      step_mem->nfsi++;
-      if (retval) { return ARK_RHSFUNC_FAIL; }
+      arkProcessError(ark_mem, ARK_INNERSTEP_FAIL, __LINE__, __func__, __FILE__,
+                      "Unable to reset the inner stepper");
+      return (ARK_INNERSTEP_FAIL);
     }
+  }
 
-    ark_mem->fn_is_current = SUNTRUE;
+  /* Evaluate the slow RHS function if needed. NOTE: we decide between calling the full RHS
+   function (if ark_mem->fn is non-NULL) versus just updating the stored value of
+   Fse[0]. In either case, we use ARK_FULLRHS_START mode because MERK methods do not
+   evaluate Fse at the end of the time step (so nothing can be leveraged). */
+  if (ark_mem->fn != NULL)
+  {
+    retval = mriStep_FullRHS(ark_mem, ark_mem->tn, ark_mem->yn, ark_mem->fn,
+                             ARK_FULLRHS_START);
+    if (retval) { return ARK_RHSFUNC_FAIL; }
```

Comment:
Skip evaluation if possible e.g., re-attempting failed step
```suggestion
    if (!ark_mem->fn_is_current)
    {
      retval = mriStep_FullRHS(ark_mem, ark_mem->tn, ark_mem->yn, ark_mem->fn,
                               ARK_FULLRHS_START);
      if (retval) { return ARK_RHSFUNC_FAIL; }
    }
```